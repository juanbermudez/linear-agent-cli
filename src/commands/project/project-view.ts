import { Command } from "@cliffy/command"
import { renderMarkdown } from "@littletof/charmd"
import { gql } from "../../__codegen__/gql.ts"
import { getGraphQLClient } from "../../utils/graphql.ts"
import { formatRelativeTime } from "../../utils/display.ts"
import { openProjectPage } from "../../utils/actions.ts"

const GetProjectDetails = gql(`
  query GetProjectDetails($id: String!) {
    project(id: $id) {
      id
      name
      description
      slugId
      icon
      color
      status {
        id
        name
        color
      }
      creator {
        name
        displayName
      }
      lead {
        name
        displayName
      }
      priority
      health
      startDate
      targetDate
      startedAt
      completedAt
      canceledAt
      updatedAt
      createdAt
      url
      teams {
        nodes {
          id
          key
          name
        }
      }
      issues {
        nodes {
          id
          identifier
          title
          state {
            name
            type
          }
        }
      }
      lastUpdate {
        id
        body
        health
        createdAt
        user {
          name
          displayName
        }
      }
    }
  }
`)

export const viewCommand = new Command()
  .name("view")
  .description("View project details")
  .alias("v")
  .arguments("<projectId:string>")
  .option("-w, --web", "Open in web browser")
  .option("-a, --app", "Open in Linear.app")
  .option("-j, --json", "Output result as JSON")
  .option("--format <format:string>", "Output format: text|json")
  .action(async (options, projectId) => {
    const { web, app, json, format } = options
    const useJson = json || format === "json"

    if (web || app) {
      await openProjectPage(projectId, { app, web: !app })
      return
    }

    const { Spinner } = await import("@std/cli/unstable-spinner")
    const showSpinner = !useJson && Deno.stdout.isTerminal()
    const spinner = showSpinner ? new Spinner() : null
    spinner?.start()

    try {
      const client = getGraphQLClient()
      const result = await client.request(GetProjectDetails, { id: projectId })
      spinner?.stop()

      const project = result.project
      if (!project) {
        const errorMsg = `Project with ID "${projectId}" not found.`
        if (useJson) {
          console.error(
            JSON.stringify({
              success: false,
              error: { code: "NOT_FOUND", message: errorMsg },
            }, null, 2),
          )
        } else {
          console.error(errorMsg)
        }
        Deno.exit(1)
      }

      if (useJson) {
        // Calculate issue summary
        const issuesByState = project.issues.nodes.reduce(
          (acc: Record<string, number>, issue) => {
            const stateType = issue.state.type
            if (!acc[stateType]) acc[stateType] = 0
            acc[stateType]++
            return acc
          },
          {} as Record<string, number>,
        )

        console.log(
          JSON.stringify({
            project: {
              id: project.id,
              name: project.name,
              slugId: project.slugId,
              description: project.description,
              url: project.url,
              icon: project.icon,
              color: project.color,
              status: {
                id: project.status.id,
                name: project.status.name,
                color: project.status.color,
              },
              creator: project.creator
                ? {
                  name: project.creator.name,
                  displayName: project.creator.displayName,
                }
                : null,
              lead: project.lead
                ? {
                  name: project.lead.name,
                  displayName: project.lead.displayName,
                }
                : null,
              priority: project.priority,
              health: project.health,
              startDate: project.startDate,
              targetDate: project.targetDate,
              startedAt: project.startedAt,
              completedAt: project.completedAt,
              canceledAt: project.canceledAt,
              createdAt: project.createdAt,
              updatedAt: project.updatedAt,
              teams: project.teams.nodes.map((t) => ({
                id: t.id,
                key: t.key,
                name: t.name,
              })),
              issuesSummary: {
                total: project.issues.nodes.length,
                byState: issuesByState,
              },
              lastUpdate: project.lastUpdate
                ? {
                  id: project.lastUpdate.id,
                  health: project.lastUpdate.health,
                  bodyPreview: project.lastUpdate.body.substring(0, 200),
                  createdAt: project.lastUpdate.createdAt,
                  author: {
                    name: project.lastUpdate.user.name,
                    displayName: project.lastUpdate.user.displayName,
                  },
                }
                : null,
            },
          }, null, 2),
        )
        return
      }

      // Build the display
      const lines: string[] = []

      // Title with icon and color
      const icon = project.icon ? `${project.icon} ` : ""
      lines.push(`# ${icon}${project.name}`)
      lines.push("")

      // Basic info
      lines.push(`**Slug:** ${project.slugId}`)
      lines.push(`**URL:** ${project.url}`)

      // Status with color styling
      const statusLine = `**Status:** ${project.status.name}`
      if (Deno.stdout.isTerminal()) {
        console.log(`%c${statusLine}%c`, `color: ${project.status.color}`, "")
      } else {
        lines.push(statusLine)
      }

      // Priority
      const priorityMap = {
        0: "None",
        1: "Urgent",
        2: "High",
        3: "Medium",
        4: "Low",
      }
      const priority =
        priorityMap[project.priority as keyof typeof priorityMap] || "None"
      lines.push(`**Priority:** ${priority}`)

      // Health
      if (project.health) {
        lines.push(`**Health:** ${project.health}`)
      }

      // People
      if (project.creator) {
        lines.push(
          `**Creator:** ${project.creator.displayName || project.creator.name}`,
        )
      }
      if (project.lead) {
        lines.push(
          `**Lead:** ${project.lead.displayName || project.lead.name}`,
        )
      }

      // Dates
      if (project.startDate) {
        lines.push(`**Start Date:** ${project.startDate}`)
      }
      if (project.targetDate) {
        lines.push(`**Target Date:** ${project.targetDate}`)
      }
      if (project.startedAt) {
        lines.push(`**Started At:** ${formatRelativeTime(project.startedAt)}`)
      }
      if (project.completedAt) {
        lines.push(
          `**Completed At:** ${formatRelativeTime(project.completedAt)}`,
        )
      }
      if (project.canceledAt) {
        lines.push(
          `**Canceled At:** ${formatRelativeTime(project.canceledAt)}`,
        )
      }

      // Teams
      if (project.teams.nodes.length > 0) {
        const teamList = project.teams.nodes
          .map((team) => `${team.name} (${team.key})`)
          .join(", ")
        lines.push(`**Teams:** ${teamList}`)
      }

      lines.push("")
      lines.push(`**Created:** ${formatRelativeTime(project.createdAt)}`)
      lines.push(`**Updated:** ${formatRelativeTime(project.updatedAt)}`)

      // Description
      if (project.description) {
        lines.push("")
        lines.push("## Description")
        lines.push("")
        lines.push(project.description)
      }

      // Latest update
      if (project.lastUpdate) {
        lines.push("")
        lines.push("## Latest Update")
        lines.push("")
        const update = project.lastUpdate
        lines.push(`**By:** ${update.user.displayName || update.user.name}`)
        lines.push(`**When:** ${formatRelativeTime(update.createdAt)}`)
        if (update.health) {
          lines.push(`**Health:** ${update.health}`)
        }
        lines.push("")
        lines.push(update.body)
      }

      // Issue summary
      if (project.issues.nodes.length > 0) {
        lines.push("")
        lines.push("## Issues Summary")
        lines.push("")

        const issuesByState = project.issues.nodes.reduce(
          (acc: Record<string, number>, issue) => {
            const stateType = issue.state.type
            if (!acc[stateType]) acc[stateType] = 0
            acc[stateType]++
            return acc
          },
          {} as Record<string, number>,
        )

        const total = project.issues.nodes.length
        const completed = issuesByState.completed || 0
        const started = issuesByState.started || 0
        const unstarted = issuesByState.unstarted || 0
        const canceled = issuesByState.canceled || 0
        const backlog = issuesByState.backlog || 0
        const triage = issuesByState.triage || 0

        lines.push(`**Total Issues:** ${total}`)
        if (completed > 0) lines.push(`**Completed:** ${completed}`)
        if (started > 0) lines.push(`**In Progress:** ${started}`)
        if (unstarted > 0) lines.push(`**To Do:** ${unstarted}`)
        if (backlog > 0) lines.push(`**Backlog:** ${backlog}`)
        if (triage > 0) lines.push(`**Triage:** ${triage}`)
        if (canceled > 0) lines.push(`**Canceled:** ${canceled}`)
      }

      const markdown = lines.join("\n")

      if (Deno.stdout.isTerminal()) {
        const terminalWidth = Deno.consoleSize().columns
        console.log(renderMarkdown(markdown, { lineWidth: terminalWidth }))
      } else {
        console.log(markdown)
      }
    } catch (error) {
      spinner?.stop()
      const errorMsg = `Failed to fetch project details: ${error}`
      if (useJson) {
        console.error(
          JSON.stringify({
            success: false,
            error: { code: "API_ERROR", message: errorMsg },
          }, null, 2),
        )
      } else {
        console.error(errorMsg)
      }
      Deno.exit(1)
    }
  })
