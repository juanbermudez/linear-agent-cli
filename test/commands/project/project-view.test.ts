import { snapshotTest } from "@cliffy/testing"
import { viewCommand } from "../../../src/commands/project/project-view.ts"
import { MockLinearServer } from "../../utils/mock_linear_server.ts"

// Common Deno args for permissions
const denoArgs = [
  "--allow-env=GITHUB_*,GH_*,LINEAR_*,NODE_ENV,EDITOR,SNAPSHOT_TEST_NAME",
  "--allow-read",
  "--allow-write",
  "--allow-run",
  "--allow-net",
  "--quiet",
]

// Test help output
await snapshotTest({
  name: "Project View Command - Help Text",
  meta: import.meta,
  colors: false,
  args: ["--help"],
  denoArgs,
  async fn() {
    await viewCommand.parse()
  },
})

// Test with mock server - Project details
await snapshotTest({
  name: "Project View Command - With Project Details",
  meta: import.meta,
  colors: false,
  args: ["project-123"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetProjectDetails",
        variables: { id: "project-123" },
        response: {
          data: {
            project: {
              id: "project-123",
              name: "Authentication System Redesign",
              description:
                "Complete overhaul of the authentication system to improve security and user experience.\n\n## Goals\n- Implement OAuth 2.0 / OpenID Connect\n- Add multi-factor authentication\n- Improve password reset flow\n- Add social login options\n\n## Technical Requirements\n- JWT tokens with proper rotation\n- Rate limiting on auth endpoints\n- Audit logging for security events\n- GDPR compliance for user data",
              slugId: "auth-redesign-2024",
              icon: "üîê",
              color: "#3b82f6",
              status: {
                id: "status-started",
                name: "In Progress",
                color: "#f59e0b",
              },
              creator: {
                name: "john.admin",
                displayName: "John Admin",
              },
              lead: {
                name: "jane.lead",
                displayName: "Jane Lead",
              },
              priority: 2,
              health: "onTrack",
              startDate: "2024-01-15",
              targetDate: "2024-04-30",
              startedAt: "2024-01-16T09:00:00Z",
              completedAt: null,
              canceledAt: null,
              updatedAt: "2024-01-25T14:30:00Z",
              createdAt: "2024-01-10T10:00:00Z",
              url: "https://linear.app/acme/project/auth-redesign-2024",
              teams: {
                nodes: [
                  {
                    id: "team-1",
                    key: "BACKEND",
                    name: "Backend Team",
                  },
                  {
                    id: "team-2",
                    key: "SECURITY",
                    name: "Security Team",
                  },
                ],
              },
              issues: {
                nodes: [
                  {
                    id: "issue-1",
                    identifier: "AUTH-101",
                    title: "Implement OAuth 2.0 flow",
                    state: {
                      name: "In Progress",
                      type: "started",
                    },
                  },
                  {
                    id: "issue-2",
                    identifier: "AUTH-102",
                    title: "Add MFA support",
                    state: {
                      name: "To Do",
                      type: "unstarted",
                    },
                  },
                  {
                    id: "issue-3",
                    identifier: "AUTH-103",
                    title: "Design new login UI",
                    state: {
                      name: "Done",
                      type: "completed",
                    },
                  },
                  {
                    id: "issue-4",
                    identifier: "AUTH-104",
                    title: "Security audit of current system",
                    state: {
                      name: "Canceled",
                      type: "canceled",
                    },
                  },
                ],
              },
              lastUpdate: {
                id: "update-1",
                body:
                  "Great progress this week! The OAuth implementation is nearly complete and we're on track for our Q1 delivery. The team has been collaborating well across backend and security concerns.\n\n**This week's highlights:**\n- OAuth 2.0 flow implementation 80% complete\n- MFA design reviews completed\n- Security penetration testing scheduled\n\n**Next week:**\n- Complete OAuth testing\n- Begin MFA implementation\n- Finalize UI designs",
                health: "onTrack",
                createdAt: "2024-01-22T16:00:00Z",
                user: {
                  name: "jane.lead",
                  displayName: "Jane Lead",
                },
              },
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})

// Test with minimal project (no optional fields)
await snapshotTest({
  name: "Project View Command - Minimal Project",
  meta: import.meta,
  colors: false,
  args: ["minimal-project"],
  denoArgs,
  async fn() {
    const server = new MockLinearServer([
      {
        queryName: "GetProjectDetails",
        variables: { id: "minimal-project" },
        response: {
          data: {
            project: {
              id: "minimal-project",
              name: "Simple Project",
              description: "",
              slugId: "simple",
              icon: null,
              color: "#64748b",
              status: {
                id: "status-backlog",
                name: "Backlog",
                color: "#94a3b8",
              },
              creator: null,
              lead: null,
              priority: 0,
              health: null,
              startDate: null,
              targetDate: null,
              startedAt: null,
              completedAt: null,
              canceledAt: null,
              updatedAt: "2024-01-20T12:00:00Z",
              createdAt: "2024-01-20T12:00:00Z",
              url: "https://linear.app/acme/project/simple",
              teams: {
                nodes: [],
              },
              issues: {
                nodes: [],
              },
              lastUpdate: null,
            },
          },
        },
      },
    ])

    try {
      await server.start()
      Deno.env.set("LINEAR_GRAPHQL_ENDPOINT", server.getEndpoint())
      Deno.env.set("LINEAR_API_KEY", "Bearer test-token")

      await viewCommand.parse()
    } finally {
      await server.stop()
      Deno.env.delete("LINEAR_GRAPHQL_ENDPOINT")
      Deno.env.delete("LINEAR_API_KEY")
    }
  },
})
