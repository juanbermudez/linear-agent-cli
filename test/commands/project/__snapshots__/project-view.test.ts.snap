export const snapshot = {};

snapshot[`Project View Command - Help Text 1`] = `
stdout:
"
Usage: view <projectId>

Description:

  View project details

Options:

  -h, --help  - Show this help.      
  -w, --web   - Open in web browser  
  -a, --app   - Open in Linear.app   

"
stderr:
""
`;

snapshot[`Project View Command - With Project Details 1`] = `
stdout:
"# üîê Authentication System Redesign

**Slug:** auth-redesign-2024
**URL:** https://linear.app/acme/project/auth-redesign-2024
**Status:** In Progress
**Priority:** High
**Health:** onTrack
**Creator:** John Admin
**Lead:** Jane Lead
**Start Date:** 2024-01-15
**Target Date:** 2024-04-30
**Started At:** 1/16/2024
**Teams:** Backend Team (BACKEND), Security Team (SECURITY)

**Created:** 1/10/2024
**Updated:** 1/25/2024

## Description

Complete overhaul of the authentication system to improve security and user experience.

## Goals
- Implement OAuth 2.0 / OpenID Connect
- Add multi-factor authentication
- Improve password reset flow
- Add social login options

## Technical Requirements
- JWT tokens with proper rotation
- Rate limiting on auth endpoints
- Audit logging for security events
- GDPR compliance for user data

## Latest Update

**By:** Jane Lead
**When:** 1/22/2024
**Health:** onTrack

Great progress this week! The OAuth implementation is nearly complete and we're on track for our Q1 delivery. The team has been collaborating well across backend and security concerns.

**This week's highlights:**
- OAuth 2.0 flow implementation 80% complete
- MFA design reviews completed
- Security penetration testing scheduled

**Next week:**
- Complete OAuth testing
- Begin MFA implementation
- Finalize UI designs

## Issues Summary

**Total Issues:** 4
**Completed:** 1
**In Progress:** 1
**To Do:** 1
**Canceled:** 1
"
stderr:
""
`;

snapshot[`Project View Command - Minimal Project 1`] = `
stdout:
"# Simple Project

**Slug:** simple
**URL:** https://linear.app/acme/project/simple
**Status:** Backlog
**Priority:** None

**Created:** 1/20/2024
**Updated:** 1/20/2024
"
stderr:
""
`;
